File: web/cockpit/chegada-detector.js
// chegada-detector.js — detecta cruzamento da linha de chegada (volta fechada).
//
// Substitui a heurística antiga "saída do último trecho = volta fechada"
// pelo sinal real: o marco de chegada cadastrado na tabela `marcos` do banco.
//
// Decisão Flávio 2026-05-28.
//
// Uso:
//   import { ChegadaDetector } from './chegada-detector.js';
//   const det = new ChegadaDetector({ marco, onChegada: () => {...} });
//   det.ingestGps({ lat, lng, t });
//
// Marco esperado:
//   { a_gps: { lat, lng }, b_gps: { lat, lng } }
//
// Algoritmo:
//   - Mesma técnica do TrechoDetector: sinal do produto vetorial muda
//     entre 2 amostras consecutivas em relação à linha → cruzou.
//   - Debounce: ignora cruzamentos consecutivos a menos de 5 segundos
//     (evita falso disparo se carro fica em pé sobre a linha).

import { SUPABASE_URL, SUPABASE_ANON_KEY } from './supabase-config.js';
import { sideOfLine, caminhoCruzaLinha } from './geo.js';

const SUPABASE_ANON = SUPABASE_ANON_KEY;

const DEBOUNCE_MS = 5000;
// Cruzamento só vale se o CAMINHO entre 2 amostras corta a linha DE VERDADE
// (segmento a–b com folga). Sem isso, o prolongamento infinito da linha corta a
// pista inteira e fecha "volta" em qualquer canto (achado no replay real, 10/06).
// A geometria (sideOfLine / caminhoCruzaLinha) mora na casa única src/domain/geo.js —
// a mesma usada pelo cérebro (chegada-gps.js). Acabou a cópia espelho.

export class ChegadaDetector {
  /**
   * @param {Object} opts
   * @param {{a_gps:{lat,lng}, b_gps:{lat,lng}}} opts.marco
   * @param {() => void} opts.onChegada  Chamado ao detectar cruzamento.
   */
  constructor({ marco, onChegada = () => {} } = {}) {
    if (!marco || !marco.a_gps || !marco.b_gps) {
      throw new Error('ChegadaDetector: marco { a_gps, b_gps } obrigatório');
    }
    this._a = marco.a_gps;
    this._b = marco.b_gps;
    this._onChegada = onChegada;
    this._lastSide = null;
    this._lastCrossT = -Infinity;
    this._voltas = 0;
  }

  ingestGps(sample) {
    if (!sample || typeof sample.lat !== 'number' || typeof sample.lng !== 'number') return;
    const t = typeof sample.t === 'number' ? sample.t : Date.now();
    const side = sideOfLine(sample, this._a, this._b);
    if (this._lastSide != null
        && Math.sign(side) !== Math.sign(this._lastSide)
        && this._lastPos
        && caminhoCruzaLinha(this._lastPos, sample, this._a, this._b)
        && t - this._lastCrossT > DEBOUNCE_MS) {
      this._lastCrossT = t;
      this._voltas++;
      try { this._onChegada({ t, voltaN: this._voltas }); }
      catch (e) { console.warn('[chegada-detector] onChegada erro:', e.message); }
    }
    this._lastSide = side;
    this._lastPos = { lat: sample.lat, lng: sample.lng };
  }

  getVoltas() { return this._voltas; }
}

/**
 * Carrega o marco de chegada de um layout no Supabase.
 */
export async function loadMarcoChegada(layoutId) {
  const { createClient } = await import('https://esm.sh/@supabase/supabase-js@2.45.0');
  const sb = createClient(SUPABASE_URL, SUPABASE_ANON);
  const { data, error } = await sb
    .from('marcos')
    .select('id, posicao, label')
    .eq('layout_id', layoutId)
    .eq('tipo', 'chegada')
    .maybeSingle();
  if (error || !data || !data.posicao) {
    console.warn('[chegada-detector] marco não encontrado:', error?.message);
    return null;
  }
  return data.posicao;
}

File: windows/cockpit/P1Fast.Cockpit.Domain/ChegadaDetector.cs
// ChegadaDetector — porte fiel de web/cockpit/chegada-detector.js (decisão Flávio
// 2026-05-28). Primeiro tijolo do MS-13.5: detecta o cruzamento da linha de chegada
// a partir do GPS — que o .exe agora lê nativo do RaceBox (ADR-024 amendment 6) —
// fechando a volta e dando o tempo de volta.
//
// Algoritmo idêntico ao JS:
//   - troca de sinal do produto vetorial (sideOfLine) entre 2 amostras consecutivas;
//   - confirmação por interseção REAL do segmento caminho × linha (folga de 50% em
//     cada ponta da linha) — sem isso o prolongamento infinito da linha fecharia
//     "volta" em qualquer canto (achado no replay real de 10/06);
//   - debounce de 5 s pra não disparar com o carro parado em cima da linha.
//
// Puro e sem IO (igual aos outros do domínio): o chamador empurra GPS e recebe o
// evento. O loader Supabase do marco (loadMarcoChegada no JS) fica fora — quem dá a
// linha é o chamador (semente da pista ou banco).

namespace P1Fast.Cockpit.Domain;

/// <summary>Ponto GPS de uma das pontas da linha de chegada.</summary>
public readonly record struct GpsPonto(double Lat, double Lng);

/// <summary>Volta fechada: instante, número da volta e o tempo da volta que fechou.</summary>
public sealed record ChegadaEvento(long TMs, int VoltaN, long? TempoVoltaMs);

public sealed class ChegadaDetector
{
    private const long DebounceMs = 5000;

    private readonly GpsPonto _a;
    private readonly GpsPonto _b;
    private double? _lastSide;
    private GpsPonto? _lastPos;
    private long _lastCrossT;
    private bool _jaCruzou;
    private int _voltas;

    /// <param name="a">Ponta A da linha de chegada.</param>
    /// <param name="b">Ponta B da linha de chegada.</param>
    public ChegadaDetector(GpsPonto a, GpsPonto b)
    {
        _a = a;
        _b = b;
    }

    public int Voltas => _voltas;

    /// <summary>
    /// Ingere uma amostra de GPS. Devolve o <see cref="ChegadaEvento"/> se fechou
    /// volta nesta amostra; senão null. TempoVoltaMs é null no primeiro cruzamento
    /// (não há volta anterior pra cronometrar).
    /// </summary>
    public ChegadaEvento? IngestGps(double lat, double lng, long tMs)
    {
        if (double.IsNaN(lat) || double.IsNaN(lng) || double.IsInfinity(lat) || double.IsInfinity(lng))
            return null;

        var p = new GpsPonto(lat, lng);
        double side = SideOfLine(p, _a, _b);

        ChegadaEvento? evento = null;
        if (_lastSide is { } ls
            && Math.Sign(side) != Math.Sign(ls)
            && _lastPos is { } lp
            && CaminhoCruzaLinha(lp, p, _a, _b)
            && (!_jaCruzou || tMs - _lastCrossT > DebounceMs))
        {
            long? tempo = _jaCruzou ? tMs - _lastCrossT : null;
            _lastCrossT = tMs;
            _jaCruzou = true;
            _voltas++;
            evento = new ChegadaEvento(tMs, _voltas, tempo);
        }

        _lastSide = side;
        _lastPos = p;
        return evento;
    }

    private static double SideOfLine(GpsPonto p, GpsPonto a, GpsPonto b) =>
        (b.Lng - a.Lng) * (p.Lat - a.Lat) - (b.Lat - a.Lat) * (p.Lng - a.Lng);

    // O caminho p0→p1 corta a LINHA a–b de verdade? Interseção de segmentos em
    // projeção local (metros), com folga de 50% em cada ponta da linha.
    private static bool CaminhoCruzaLinha(GpsPonto p0, GpsPonto p1, GpsPonto a, GpsPonto b)
    {
        const double kLat = 110540;
        double kLng = 111320 * Math.Cos(a.Lat * Math.PI / 180);
        double X(GpsPonto q) => (q.Lng - a.Lng) * kLng;
        double Y(GpsPonto q) => (q.Lat - a.Lat) * kLat;

        double rx = X(p1) - X(p0), ry = Y(p1) - Y(p0);
        double dx = X(b),          dy = Y(b);
        double den = rx * dy - ry * dx;
        if (Math.Abs(den) < 1e-9) return false; // paralelos

        double qpx = -X(p0), qpy = -Y(p0);
        double v = (qpx * dy - qpy * dx) / den;  // posição ao longo do caminho
        double u = (qpx * ry - qpy * rx) / den;  // posição ao longo da linha
        return v >= 0 && v <= 1 && u >= -0.5 && u <= 1.5;
    }
}
